File: zenodo-api.R
library(httr2)
library(purrr)
library(fs)

global_id <- "335741" #first version actually

base_req <- request("https://sandbox.zenodo.org/api/deposit/depositions") |>
  req_auth_bearer_token(Sys.getenv("ZENODO_SANDBOX_TOKEN"))

# 1. Get id of most recent version of global_id

resp <- base_req |>
  req_url_path_append(global_id) |>
  req_perform()
record <- resp_body_json(resp)
latest <- request(record$links$latest) |>
  req_auth_bearer_token(Sys.getenv("ZENODO_SANDBOX_TOKEN")) |>
  req_perform() |>
  resp_body_json()
latest_id <- latest$id


# 2. Create new version

new_version_req <- base_req |>
  req_url_path_append(latest_id, "actions/newversion") |>
  req_method("POST")
new_version <- req_perform(new_version_req) |>
  resp_body_json()

# 3. Get id of new version from links$latest_draft

new_version_id <- request(new_version$links$latest_draft) |>
  req_auth_bearer_token(Sys.getenv("ZENODO_SANDBOX_TOKEN")) |>
  req_perform() |>
  resp_body_json() |>
  purrr::pluck("id")

# 4. Update metadata

new_metadata <- new_version$metadata
new_metadata$publication_date <- Sys.Date()

base_req |>
  req_url_path_append(new_version_id) |>
  req_method("PUT") |>
  req_body_json(data = list(metadata = new_metadata)) |>
  req_perform()

# 4. list files

file_ids <- base_req |>
  req_url_path_append(new_version_id, "files") |>
  req_perform() |>
  resp_body_json() |>
  map_chr("id")

# 5. delete files
if (length(file_ids) > 0) {
  map(file_ids, \(x) {
    base_req |>
      req_url_path_append(new_version_id, "files", x) |>
      req_method("DELETE")
  }) |>
    req_perform_sequential()
}

# 6. upload new files (using new file upload API)

# First get the bucket link
info <- base_req |>
  req_url_path_append(new_version_id) |>
  req_perform() |>
  resp_body_json()
bucket <- info$links$bucket
paths <- fs::dir_ls("fia/parquet")

# Then upload all the files
upload_reqs <- map(paths, \(path) {
  request(bucket) |>
    req_url_path_append(path_file(path)) |>
    req_auth_bearer_token(Sys.getenv("ZENODO_SANDBOX_TOKEN")) |>
    req_method("PUT") |>
    req_body_file(path)
})

# req_perform_parallel() doesn't work for some reason
req_perform_sequential(upload_reqs)

# 7. publish release
# TODO: consider making this manual so Margaret has to double-check the release
base_req |>
  req_url_path_append(new_version_id, "actions", "publish") |>
  req_method("POST") |>
  req_perform()

File: zenodo-api.py
import requests
from pathlib import Path
import os
import datetime
import json
import time
import shutil as sh
import subprocess as sp
# make new folders that hold the 4 parquet files that are available
parquets = sorted(Path("fia/parquet").glob("*.parquet"))

# make the folders
for p in parquets:
  # the first part of the name separated by underscores is the state part

  state_name = p.name.split("_")[0]
  folder = Path(f"fia/parquet/{state_name}")
  folder.mkdir(exist_ok=True)
  # try to move the file into that folder
  sh.move(p,f"{folder}/{p.name}")
  print("moved",p)

new_locations = sorted(Path("fia/parquet").glob("*/*.parquet"))
print(new_locations)

# now run a system command to zip up the contents of the state folders
folders = sorted(Path("fia/parquet").glob("*"))

for folder in folders:
  sp.run(f"zip -r fia/parquet/{folder.name}.zip {folder}",shell=True)
  print("folder zipped",folder)




ACCESS_TOKEN=os.environ["ZENODO_TOKEN"]

# try to list out the existing r

conceptid = "17088642"
headers = {"Authorization":f"Bearer {ACCESS_TOKEN}"}

r= requests.get("https://zenodo.org/api/deposit/depositions",params={"status":"draft","q":conceptid},headers=headers)
print(r.json())

# go through list and remove any  drafts, new versions or otherwise
resultj = r.json()

for res in resultj:
  print("removing",res)
  draft_id = res["id"]
  r= requests.delete(f"https://zenodo.org/api/deposit/depositions/{draft_id}",headers=headers)
  print()


# get the latest version
#gid = "20635476"
gid = "17088643"
r = requests.get(f"https://zenodo.org/api/deposit/depositions/{gid}",headers=headers)
# 401
res = r.json()
res["links"]["latest"]
r = requests.get( res["links"]["latest_draft"] ,headers=headers)
res = r.json()
print(res)
print(res["id"])
latest_id = res["id"]

# make a new version


headers = {
    "Content-Type": "application/json",
    "Authorization": f"Bearer {ACCESS_TOKEN}"
}
print(latest_id)
r = requests.post(f'https://zenodo.org/api/deposit/depositions/{latest_id}/actions/newversion',
                   headers=headers)

# update the metadata

print("updating metadata")
new_version_res = r.json()
print(new_version_res)
meta = new_version_res["metadata"]
x = datetime.datetime.now()
meta["publication_date"]= x.strftime("%Y-%m-%d")
meta["version"] = meta["publication_date"] 


headers = {
    "Content-Type": "application/json",
    "Authorization": f"Bearer {ACCESS_TOKEN}"
}
print(f'https://zenodo.org/api/deposit/depositions/{new_version_res["id"]}')
r = requests.put(f'https://zenodo.org/api/deposit/depositions/{new_version_res["id"]}',
                   data=json.dumps({"metadata":meta}),
                   headers=headers)




# delete the files in the new version draft

print("trying to delete contents of previous version")
r = requests.get(f'https://zenodo.org/api/deposit/depositions/{new_version_res["id"]}/files',headers=headers)
for f in r.json():
    try:
      r = requests.delete(f'https://zenodo.org/api/deposit/depositions/{new_version_res["id"]}/files/{f["id"]}',headers=headers)

    except Exception as e:
      print("ran into ")
      print(e)
      print()
      print("continuing")

# adding new contents
print("adding new contents")
r = requests.get(f'https://zenodo.org/api/deposit/depositions/{new_version_res["id"]}',headers=headers)
print(r)
new_bucket_url = r.json()["links"]["bucket"]
''' 
The target URL is a combination of the bucket link with the desired filename
seperated by a slash.
'''
files = sorted(Path("fia/parquet/").glob("*.zip"))
print(files)
headers = {
    "Content-Type": "application/octet-stream",
    "Authorization": f"Bearer {ACCESS_TOKEN}"
}
for f in files:
  print("starting",f)
  with open(f, "rb") as fp:
    try:
      r = requests.put(
          "%s/%s" % (new_bucket_url,f.name),
          data=fp,
          headers=headers,
      )
      print(r)
      print(r.json())
    except Exception as e:
      print("upload snag")
      print(e)
      print("retrying after delay")
      failed = True
      retries = 5
      while failed:
        retries -=1
        if retries <0:
          print("failed completely",f)
          break
        time.sleep(2)
        try:
          r = requests.put(
              "%s/%s" % (new_bucket_url,f.name),
              data=fp,
              headers=headers,
          )
          print(r)
          failed = False
        except:
          print("failed upload again",f)

          

    print("uploaded ",f)
r.json()

headers = {'Authorization': f'Bearer {ACCESS_TOKEN}'}
r = requests.post('https://zenodo.org/api/deposit/depositions/%s/actions/publish' % new_version_res["id"],
                      headers=headers)
r.status_code